spec.html

```diff
@@ -40393,6 +40496,90 @@ <h1>`Promise.allSettled` Reject Element Functions</h1>
+          1. If _result_ is an abrupt completion, then
+            1. If _iteratorRecord_.[[Done]] is *false*, set _result_ to IteratorClose(_iteratorRecord_, _result_).
+            1. IfAbruptRejectPromise(_result_, _promiseCapability_).
+          1. Return Completion(_result_).
+        </emu-alg>
+        <emu-note>
+          <p>The `any` function requires its *this* value to be a constructor function that supports the parameter conventions of the `Promise` constructor.</p>
+        </emu-note>
+
+        <emu-clause id="sec-performpromiseany" aoid="PerformPromiseAny">
+          <h1>Runtime Semantics: PerformPromiseAny ( _iteratorRecord_, _constructor_, _resultCapability_, _promiseResolve_ )</h1>
+          <p>The abstract operation PerformPromiseAny takes arguments _iteratorRecord_, _constructor_, _resultCapability_ (a PromiseCapability Record), and _promiseResolve_. It performs the following steps when called:</p>
+          <emu-alg>
+            1. Assert: ! IsConstructor(_constructor_) is *true*.
+            1. Assert: ! IsCallable(_promiseResolve_) is *true*.
+            1. Let _errors_ be a new empty List.
+            1. Let _remainingElementsCount_ be a new Record { [[Value]]: 1 }.
+            1. Let _index_ be 0.
+            1. Repeat,
+              1. Let _next_ be IteratorStep(_iteratorRecord_).
+              1. If _next_ is an abrupt completion, set _iteratorRecord_.[[Done]] to *true*.
+              1. ReturnIfAbrupt(_next_).
+              1. If _next_ is *false*, then
+                1. Set _iteratorRecord_.[[Done]] to *true*.
+                1. Set _remainingElementsCount_.[[Value]] to _remainingElementsCount_.[[Value]] - 1.
+                1. If _remainingElementsCount_.[[Value]] is 0, then
+                  1. Let _error_ be a newly created `AggregateError` object.
+                  1. Perform ! DefinePropertyOrThrow(_error_, `"errors"`, Property Descriptor { [[Configurable]]: *true*, [[Enumerable]]: *false*, [[Writable]]: *true*, [[Value]]: ! CreateArrayFromList(_errors_) }).
+                  1. Return ThrowCompletion(_error_).
+              1. Return _resultCapability_.[[Promise]].
+            1. Let _nextValue_ be IteratorValue(_next_).
+            1. If _nextValue_ is an abrupt completion, set _iteratorRecord_.[[Done]] to *true*.
+            1. ReturnIfAbrupt(_nextValue_).
+            1. Append *undefined* to _errors_.
+            1. Let _nextPromise_ be ? Call(_promiseResolve_, _constructor_, &laquo; _nextValue_ &raquo;).
+            1. Let _steps_ be the algorithm steps defined in <emu-xref href="#sec-promise.any-reject-element-functions" title></emu-xref>.
+            1. Let _rejectElement_ be ! CreateBuiltinFunction(_steps_, &laquo; [[AlreadyCalled]], [[Index]], [[Errors]], [[Capability]], [[RemainingElements]] &raquo;).
+            1. Set _rejectElement_.[[AlreadyCalled]] to a new Record { [[Value]]: *false* }.
```

Comment:
It is not totally clear to me why this needs to be a Record. I guess that's precedent in `Promise.all`, but I don't know why that uses a Record either.

Reply:
Yes, precedents are in `Promise.all` and `Promise.allSettled`

Reply:
I think we probably want to fix this, but not in this PR.

Reply:
Afaics it has to be a Record (instead of just a bool) for allSettled. In that case, both the resolve builtin and the reject builtin point to the same Record. So that if the resolve builtin runs, and then the reject builtin runs, the latter has to do early return.

@syg can prob confirm this :) (We just bumped into this in a bug...)

-> And I'd assume it's a Record for all combinators for consistency. Or were you planning to keep it a Record for allSettled and make it a bool for others?

Reply:
@bakkot and I discussed this a bit on the editor call last week. The Record is really needed only for `allSettled`, which has **both** a fulfill and a reject closure per element. All other combinators have a single fulfill closure per element, making the Record unnecessary.

The pro of keeping it a Record is that all the combinators then have the same machinery.

The pro of making it a single boolean for everything but `allSettled` is then people don't need to wonder "why is this a Record", but the con is they might naively assume `allSettled` uses the same machinery and then get it wrong.

Reply:
`Promise.all()` was changed to a Record type per <https://tc39.es/archives/bugzilla/2636/>, because at that point of time we still had `Function.prototype.toMethod()`.